dask/compose.py

```diff
@@ -0,0 +1,299 @@
+import operator
+from functools import partial, wraps
+from itertools import chain
+from collections import Iterator
+
+from toolz import merge, unique
+
+from .core import istask
+from .optimize import cull, fuse
+from .context import _globals
+from .compatibility import apply
+from . import threaded
+
+__all__ = ['do', 'value', 'Value']
+
+
+def flat_unique(ls):
+    """Flatten ``ls``, filter by unique id, and return a list"""
+    return list(unique(chain.from_iterable(ls), key=id))
+
+
+def unzip(ls, nout):
+    """Unzip a list of lists into ``nout`` outputs."""
+    out = list(zip(*ls))
+    if not out:
+        out = [()] * nout
+    return out
+
+
+def to_task_dasks(task):
+    """Normalize a task and extract all sub-dasks.
+
+    - Replace ``Values`` with their keys
+    - Convert literals to things the schedulers can handle
+    - Extract dasks from all enclosed values
+
+    Returns
+    -------
+    task : normalized task to be run
+    dasks : list of dasks that form the dag for this task
+
+    Examples
+    --------
+
+    >>> a = value(1, 'a')
+    >>> b = value(2, 'b')
+    >>> task, dasks = to_task_dasks([a, b, 3])
+    >>> task # doctest: +SKIP
+    (list, ['a', 'b', 3])
+    >>> dasks # doctest: +SKIP
+    [{'a': 1}, {'b': 2}]
+
+    >>> task, dasks = to_task_dasks({a: 1, b: 2})
+    >>> task # doctest: +SKIP
+    (dict, (list, [(list, ['a', 1]), (list, ['b', 2])]))
+    >>> dasks # doctest: +SKIP
+    [{'a': 1}, {'b': 2}]
+    """
+    if istask(task) and not isinstance(task[0], Value):
+        args, dasks = unzip(map(to_task_dasks, task[1:]), 2)
+        return (task[0],) + tuple(args), flat_unique(dasks)
```

Comment:
I was surprised to see that these lines weren't covered.

(sorry to be a bit picky on this PR by the way.  I expect this code to be used by non-expert users a fair amount.)